File: examples/rockset/golang/basic.go
/*
Example program using the official Rockset Golang Client with a CrateDB backend.
Here: "Documents" and "Queries" APIs, basic usage.

Usage
=====
- Adjust configuration: http://localhost:4243
- Run program: go run basic.go

Documentation
=============
- https://pkg.go.dev/github.com/rockset/rockset-go-client#RockClient.AddDocuments
- https://pkg.go.dev/github.com/rockset/rockset-go-client#RockClient.Query
- https://docs.rockset.com/documentation/reference/adddocuments
- https://docs.rockset.com/documentation/reference/query
*/

package main

import (
	"context"
	"fmt"
	"os"
    "time"

	"github.com/rs/zerolog"

    "github.com/rockset/rockset-go-client"
)


func main() {

    console := zerolog.ConsoleWriter{Out: os.Stderr, TimeFormat: time.RFC3339}
    log := zerolog.New(console).Level(zerolog.TraceLevel).With().Timestamp().Logger()
    ctx := log.WithContext(context.Background())

    // The Golang client does not allow setting the scheme from the URL as it only supports HTTPS.
    // https://github.com/rockset/rockset-go-client/blob/v0.24.2/rockset.go#L114
    // => Error: http: server gave HTTP response to HTTPS client
    rc, err := rockset.NewClient(rockset.WithAPIKey("abc123"), rockset.WithAPIServer("http://localhost:4243"))

    // rc.cfg undefined (type *rockset.RockClient has no field or method cfg)
    //rc.cfg.Scheme = "http://"

    fmt.Println("rc:", rc)
    fmt.Println("err:", err)

    /*
    response, err := rs.AddDocumentsWithOffset(ctx, "commons", "users", docs)
    if err != nil {
        return err
    }
    w := wait.New(rs)
    err = w.UntilQueryable(ctx, "commons", "users", []string{response.GetLastOffset()})
    */

    result, err := rc.Query(ctx, "SELECT * FROM commons.foobar")
    fmt.Println(result)
    fmt.Println(err)


    /*
    q := rc.QueriesApi.Query(ctx)
    rq := openapi.NewQueryRequestWithDefaults()

    rq.Sql = openapi.QueryRequestSql{Query: "SELECT * FROM commons._events where label = :label"}
    rq.Sql.DefaultRowLimit = openapi.PtrInt32(10)

    rq.Sql.Parameters = []openapi.QueryParameter{
        {
            Name:  "label",
            Type:  "string",
            Value: "QUERY_SUCCESS",
        },
    }

    r, _, err := q.Body(*rq).Execute()
    if err != nil {
        log.Fatal(err)
    }

    for _, c := range r.Collections {
        fmt.Printf("collection: %s\n", c)
    }
    */

}

File: examples/rockset/javascript/basic.js
/*
Example program using the official Rockset Javascript/Typescript Client with a CrateDB backend.
Here: "Documents" and "Queries" APIs, basic usage.

Usage
=====
- Install dependencies: Run `npm install`.
- Adjust configuration: Define `ROCKSET_APISERVER` environment variable.
- Run program: Run `npm run basic`.

Documentation
=============
- https://github.com/rockset/rockset-js
- https://github.com/rockset/rockset-js/tree/master/packages/client
- https://docs.rockset.com/documentation/reference/adddocuments
- https://docs.rockset.com/documentation/reference/query
*/
import rockset from '@rockset/client';

// Connect to API endpoint.
const rockset_apiserver = process.env.ROCKSET_APISERVER || "http://127.0.0.1:4243";
const rockset_apikey = process.env.ROCKSET_APIKEY || "abc123";
const rocksetClient = rockset.default(rockset_apikey, rockset_apiserver);

// Add documents into your collection, using the Write API.
// https://github.com/rockset/rockset-js/tree/master/packages/client#add-documents-into-your-collection-with-the-write-api
rocksetClient.documents
  .addDocuments(
    'commons' /* name of workspace */,
    'users' /* name of collection */,
    {
      /* array of JSON objects, each representing a new document */
      data: [
        { first_name: 'Brian', last_name: 'Morris', age: '14' },
        { first_name: 'Justin', last_name: 'Smith', age: '78' },
        { first_name: 'Scott', last_name: 'Wong', age: '42' },
      ],
    }
  )
  .then(console.log)
  .catch(console.error);

// Submit SQL query.
// https://github.com/rockset/rockset-js/tree/master/packages/client#run-a-sql-query
rocksetClient.queries
  .query({
    sql: {
      query: 'SELECT * FROM commons.users u WHERE u.age > :minimum_age',
      /* (optional) list of parameters that may be used in the query */
      parameters: [
        {
          name: 'minimum_age' /* name of parameter */,
          type: 'int' /* one of: int, float, bool, string date, datetime, time, timestamp */,
          value: '20' /* value of parameter */,
        },
      ],
      default_row_limit: 150 /* (optional) row limit to be used if no limit is specified in the query */,
    },
  })
  .then(console.log)
  .catch(console.error);
